src/libstd/keyword_docs.rs

```diff
@@ -1578,12 +1578,255 @@ mod type_keyword {}
+/// referred to as [soundness]: a well-typed program actually has the desired
+/// properties. In Safe Rust, the language and the compiler do this mostly by
+/// themselves, whereas in Unsafe Rust the programmer must help. The
+/// [Nomicon][nomicon-soundness] has a more detailed explanation on the subject.
+///
+/// You may have noted all uses of `unsafe` are not equivalent: some are here to
+/// note the existence of a contract the programmer must check, others are to
+/// say "I have checked the contract, go ahead and do this".
+///
+/// ## The different meanings of `unsafe`
+///
+/// Depending on the place where it is used, the `unsafe` keyword meaning can
+/// change. See the following [discussion on Rust Internals] for more in-depth
+/// explanations.
+///
+/// Summarizing the main points:
+///
+/// - `unsafe fn`: calling this function means abiding by a contract the
+/// compiler cannot enforce.
+/// - `unsafe trait`: implementing the [`trait`] means abiding by a
+/// contract the compiler cannot enforce.
+/// - `unsafe {}`: the contract necessary to call the operations inside the
+/// block has been checked and is guaranteed to be respected.
+/// - `unsafe impl`: the contract necessary to implement the trait has been
+/// checked and is guaranteed to be respected.
+///
+/// As of now (Rust 1.44), `unsafe fn` also acts like an `unsafe {}` block
+/// around the code inside the function. This means it is not just a signal to
+/// the caller, but also promises that the preconditions for the operations
+/// inside the function are upheld. Mixing these two meanings can be confusing
+/// and [proposal]s exist to use `unsafe {}` blocks inside such functions when
+/// making `unsafe` operations.
+///
+/// See the [Rustnomicon] and the [Reference] for more informations.
+///
+/// # Examples
+///
+/// The following sections are examples of the different uses of `unsafe`, often
+/// with links to go further.
+///
+/// ## Marking elements as `unsafe`
+///
+/// `unsafe` can be used on functions. Note that functions declared in
+/// [`extern`] blocks are implicitly marked as `unsafe`. Methods can also be
+/// declared as `unsafe`:
+///
+/// ```rust
+/// # #![allow(dead_code)]
+/// unsafe fn unsafe_fn() {}
+///
+/// extern "C" {
+///     fn unsafe_extern_fn();
+/// }
+///
+/// trait TraitWithUnsafeMethod {
+///     unsafe fn unsafe_method();
+/// }
+///
+/// struct S;
+///
+/// impl S {
+///     unsafe fn unsafe_method_on_struct() {}
+/// }
+/// ```
+///
+/// Traits can also be declared as `unsafe`:
+///
+/// ```rust
+/// unsafe trait UnsafeTrait {}
+/// ```
+///
+/// Documenting public `unsafe fn` (and other elements) is important. The
+/// standard library has many examples of this, like the following which is an
+/// extract from [`Vec::set_len`]. The `# Safety` section explains the contract
+/// that must be fulfilled to safely call the function. If necessary, a
+/// `# Panics` section can be added to document possible panics and their
```

Comment:
"panics" are entirely unrelated to `unsafe`, why do you even mention that here?

Reply:
Isn't it more likely that unsafe functions panics than having safe functions panics?

Reply:
I don't think so. If anything it is less likely -- unsafe functions can requrie the caller to ensure that no error conditions occur; safe functions have no choice but to panic when an unexpected error happens.

Reply:
I simply removed this sentence in the upcoming commit